Fix problem with up-to-date check and custom items

panopticoncentral opened the pull request:
**Customer scenario**

The customer adds a XAML file to their SDK project. From that point on, the up to date check believes the project is never up to date.

(This also improves the up to date check logging so it is more obvious why a check failed.)

**Bugs this fixes:** 

#2763

**Workarounds, if any**

Remove the XAML file.

**Risk**

Low. We're avoiding a previously unhandled exception.

**Performance impact**

Improves performance, avoids unhandled exception.

**Is this a regression from a previous update?**

No

**How was the bug found?**

Customer report

Pilchie commented:
Is the issue here that `FullName` isn't in the rule for Xaml items?  Also, should we just be looking at `Identity` instead of `FullName`?

sharwell commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs:
```diff
@@ -339,28 +344,28 @@ private bool CheckGlobalConditions(BuildAction buildAction, BuildUpToDateCheckLo
             if (copyAlwaysItem.Path != null)
             {
-                logger.Info("Item '{0}' has CopyToOutputDirectory set to 'Always', skipping check.", copyAlwaysItem.Path);
+                logger.Info("Item '{0}' has CopyToOutputDirectory set to 'Always', failed.", copyAlwaysItem.Path);
```
:bulb: ~~failed~~ treating as not up-to-date

sharwell commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs:
```diff
@@ -236,7 +239,9 @@ private void OnSourceItemChanged(IProjectSubscriptionUpdate e, IProjectItemSchem
-                var items = itemType.Value.After.Items.Select(item => (item.Value[FullPath], GetLink(item.Value), GetCopyType(item.Value)));
+                var items = itemType.Value.After.Items
+                    .Select(item => (GetFullPath(item.Value), GetLink(item.Value), GetCopyType(item.Value)))
+                    .Where(tuple => tuple.Item1 != null);
```
❔ Can you briefly explain how this was interfering with the XAML check, and why it's better than the workaround I was using previously?

panopticoncentral commented:
@Pilchie @sharwell Sorry, I wasn't clear. The repro project has a XAML rule file declared in its project file as a XamlPropertyRule item. It's showing up in the list of source items, but we don't have a rule for it. So the up to date check is crashing. This change basically hardens us against source items that we don't know anything about.

panopticoncentral commented:
Yeah, it's been long enough that I can't remember why I used FullPath instead of Identity, I do believe I considered it. Off the top of my head, maybe identity doesn't always include the full path? It's a little hazy. I can make a note to go back and look at it.

sharwell commented:
@panopticoncentral By the title, I was expecting this pull request to correct this: dotnet/roslyn-sdk#99

Is that not the case?

panopticoncentral commented:
Sorry, my title was misleading. It does not fix those issues.

panopticoncentral commented:
@dotnet/project-system Can someone review?

Pilchie commented:
I'd like to know if we could just switch to `Identity` instead of `FullPath` all together.  @davkean will a dataflow rule always have `Identity`?

lifengl commented:
Can you use item.Key, when item.Value[FullPath] does not exist.

davkean commented:
You'll need to resolve it via the project `UnconfiguredProject.MakeRooted`.

lifengl commented:
That will be the evaluated path of each item in the msbuild.  Frankly, I didn't realize you are using FullPath here, and I am thinking whether we can get rid of it in the rule subscription.  That property is lots of long strings in the memory, and msbuild evaluates it in a lazy way, so the property in the rule means we recreate the string on every reevaluation.

panopticoncentral commented:
@lifengl Does this look better? I moved all uses of FullPath to just use item.Key + MakeRooted. I also removed the properties on items that I had added it to since we don't need it.

panopticoncentral commented:
@dotnet-bot retest this please
